security/advancedtls/advancedtls_test.go

```diff
@@ -46,14 +47,249 @@ func Test(t *testing.T) {
 	grpctest.RunSubTests(t, s{})
 }
 
+type fakeRootProvider struct {
+	isClient  bool
+	wantError bool
+}
+
+func (fake fakeRootProvider) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	if fake.wantError {
+		return nil, fmt.Errorf("bad fakeRootProvider")
+	}
+	cs := &certStore{}
```

Comment:
Maybe move all this initialization to a `newFakeProvider` method, which returns `(*fakeProvider, error)`

Reply:
That's a great idea, but having an initialization makes it difficult to create temporary object and feed it directly to the test structs below. We can refactor the test code but that makes the test structs contains a bit more fields, since we need to include all the fields in fakeProvider to the test struct. 
As an alternative, I am considering having an `init()` function that will load all the test data into a global variable. Then this `newFakeProvider` can just return `*fakeProvider`  without error. This might go into a separate PR. What do you think about this idea?